manpages/manpages/shinken-poller.8
.\" Man page generated from reStructuredText.
.
.TH SHINKEN-POLLER 8 "2014-04-24" "2.0.1" "Shinken commands"
.SH NAME
shinken-poller \- Shinken poller daemon
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
\fBshinken\-poller\fP [\-dr] [\-c \fICONFIGFILE\fP] [\-\-debugfile \fIDEBUGFILE\fP]
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
Shinken poller daemon.
.sp
The \fBshinken\-poller\fP daemon is in charge of launching plugins as requested by schedulers. When the check is finished it returns the result to the schedulers.
.SH OPTIONS
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \-v\fP,\fB  \-\-verify\-config
Verify config file and exit
.TP
.BI \-c \ INI\-CONFIG\-FILE\fP,\fB \ \-\-config\fB= INI\-CONFIG\-FILE
Config file
.TP
.B \-d\fP,\fB  \-\-daemon
Run in daemon mode
.TP
.B \-r\fP,\fB  \-\-replace
Replace previous running poller
.TP
.B \-h\fP,\fB  \-\-help
Show this help message
.TP
.B \-\-version
Show program\(aqs version number
.TP
.BI \-\-debugfile\fB= DEBUGFILE
Enable debug logging to \fIDEBUGFILE\fP
.TP
.BI \-p \ PROFILE\fP,\fB \ \-\-profile\fB= PROFILE
Dump a profile file. Need the python cProfile library
.UNINDENT
.UNINDENT
.UNINDENT
.SH AUTHOR
Michael Leinartas,
Arthur Gautier,
David Hannequin,
Thibault Cohen
.\" Generated by docutils manpage writer.
.


Changelog
########################
Shinken ChangeLog
########################


3.0.03 - X/X/2017
-----------------
        CORE ENHANCEMENT
        Enh: (backport from Enterprise, internal ref #SEF-202) Big boost for the broker daemons, especially for very large setups and slow modules (yes Graphite, I'm looking at you).
        Enh: (backport from Enterprise ref:#SEF-103) Slight boost of startup time by remiving useless hash computation.

        CORE FIXES
        Fix: (backport from Enterprise ref:#SEF-76) In some case, we can have "negative" value for downtime depth, and then the element will never exit from downtime.


2.4.3 - 10/03/2016
-------------------
        CORE ENHANCEMENT
        Add: First version for a online profiling system, not enabled currently

        CORE FIXES
        Fix: Fix missing Services/Hosts in Livestatus (Close #1691)
        Fix: initial_state mapping for hosts
        Fix: Fix duration when a parent is down (Close #1779)
        Add: Better logging for regenerator

        MISC ENHANCEMENT
        Enh: HTML mail notifications fix and improvements
        Enh: Change TCP Connect scan to TCP Syn scan with nmap for discovery features
        Enh: Various Documentation 

2.4.2 - 01/10/2015
-------------------
        CORE ENHANCEMENT
        Add: Service excludes/includes/overrides extension
        Add: Implementation of inital_state
        Add: Broks when host/service downtime is scheduled
        Add: Solaris SMF manifests
        Enh: Allow white space between foreach elements
        Enh: Increase default http thread_pool to 16

        CORE FIXES
        Fix: Duplicate Service from template using duplicate_foreach with the same name
        Fix: Service_(includes|excludes) template recursion
        Fix: Service description with multi level inheritance 
        Fix: Default business rule notification options

2.4.1 - 15/07/2015
-------------------
        CORE ENHANCEMENT
        Add: Safe Pickle
        Add: Better Debian 8 Jessie support

        CORE FIXES
        Fix: Display_name when using duplicate_foreach
        Fix: template definition loop did segfault python
        Fix: Service Description inheritance when using several level of inheritance
        Fix: cpu looping for receiver

2.4   - 04/05/2015
-------------------
	CORE ENHANCEMENT
	Add: Hosts - Service_includes feature. Similar to service_excludes
	Add: Deprecation in the doc about the discovery part. Should be move to contrib or as a module in next versions
	Add: (by Ddurieux) better logging of failed external modules


	Enh: Actions - Catch stderr output
	Enh: Arbiter -  Unpickle new conf earlier. Prevent a memory issue
	Enh: Config - Remove sudo from restart/reload commands
	Enh: Daemon - Factorization of  get_objects_from_from_queues
	Enh: DaemonsLinks : Move files to object directory
	Enh: Datamanager - Function get_contactgroup
	Enh: Doc - A lot of typo and improvements!
	Enh: Init Scripts - use readlink to avoid failure on symlinks
	Enh: Modulesmanager - Reworking module loading, now try to import in 3 different ways
	Enh: Pep8 : Shinken is now pep8 compliant (enforced by Travis) excepted for 3 specific rules
	Enh: Scheduler - Use iterator instead of list when necessary. Reduce memory consumption
	Enh: Setup - Refactoring to be virtualenv compatible
	Enh: Tests - Stabilization


	Del: Check Shinken script. It has its own repository
	Del: Windows files from 1.4 version


	CORE FIXES
	Fix: Config Parsing - Arbiter module auto creation
	Fix: Config Parsing - Infinite recursion loop in template
	Fix: Config Parsing - Inheritance of custom variable
	Fix: Config Parsing - Raise error properly in ArbiterLink object
	Fix: Config Parsing - Type properties in Contact, NotificationWay and Arbiter objects
	Fix: Escalations - escalated parameter in notification object and macro $NOTIFICATIONISESCALATED$ are set correctly
	Fix: Event Handlers - Global event handlers parsing and launching
	Fix: External Command - Internal argument for CHANGE_RETRY_HOST_CHECK_INTERVAL command
	Fix: Realm - Bailout when there are two default realm
	Fix: (reported by:Stéphane Loeuillet) init.d - Unset http(s)_proxy env variables in the init.d script (curl was using them by default)
	Fix: (reported by: Arthur Lutz) CLI - manage shinken install call without arguments
	Fix: (reported by mohierf) STATS - fix statsd format.
	Fix: (reported by: Stéphane Loeuillet) Config Parsing - manage , case at the end of members list (hostgroups & contactgroups)
	Fix: (reported by: Stéphane Loeuillet) Config Parsing - host check_period was mandatory with some parameters, now is missing always true period.
	Fix: + char in servicegroups




2.2   - 16/01/2015
-------------------
	CORE ENHANCEMENT
	Add : Bottle - HTTP_X_FORWARDED_PROTO support (for WebUI)
	Add : BpRules - Expands bp_rules t flag from tags
	Add : BpRules - Use standard macros expansion in bp_rules output
	Add : Broks - new type unknown_check_result_brok (included in the previous add)
	Add : CLI - update command for the shinken.io install
	Add : Contact - new parameter expert. For UI purpose
	Add : Escalations - Time properties for host and service escalations
	Add : External Command - Implement SCHEDULE_HOSTGROUP_HOST_DOWNTIME and SCHEDULE_HOSTGROUP_SVC_DOWNTIME
	Add : External Commands - Reload-config function
	Add : External Commands - support for reset modify attribute
	Add : HTTP API - get_start_time call returning timestamp
	Add : HTTP API - Inner stats can be sent to shinken.io
	Add : LogEvent - Output key to properties dict
	Add : LogEvent - Pattern for FLAPPING events
	Add : Receiver - Accept passive results from host or service not in configuration
	Add : Realms - Handle properly multi level of realms
	Add : SchedulingItem - callbacks for HOSTDOWNTIME and SERVICEDOWNTIME macros
	Add : SchedulingItem - Parameter to remove host-service dependency
	Add : Service Escalation - wildcard service_description
	Add : Service - method  get_service_tags (for WebUI)

src/shinken-poller.py
#!/usr/bin/env python
#Copyright (C) 2009-2010 : 
#    Gabes Jean, naparuba@gmail.com 
#    Gerhard Lausser, Gerhard.Lausser@consol.de
#
#This file is part of Shinken.
#
#Shinken is free software: you can redistribute it and/or modify
#it under the terms of the GNU Affero General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.
#
#Shinken is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU Affero General Public License for more details.
#
#You should have received a copy of the GNU Affero General Public License
#along with Shinken.  If not, see <http://www.gnu.org/licenses/>.


#This class is an application for launch checks
#The poller listen configuration from Arbiter in a port (first argument)
#the configuration gived by arbiter is schedulers where actionner will
#take checks.
#When already launch and have a conf, poller still listen to arbiter 
#(one a timeout) if arbiter whant it to have a new conf, poller forgot
#old cheduler (and checks into) take new ones and do the (new) job.

import sys, os
import getopt
import ConfigParser
import platform

#We know that a Python 2.5 or Python3K will fail.
#We can say why and quit.
python_version = platform.python_version_tuple()

## Make sure people are using Python 2.5 or higher
if int(python_version[0]) == 2 and int(python_version[1]) < 6:
    print "Shinken require as a minimum Python 2.6.x, sorry"
    sys.exit(1)

if int(python_version[0]) == 3:
    print "Shinken is not yet compatible with Python3k, sorry"
    sys.exit(1)


try:
    import Pyro.core
except ImportError:
    print "Shinken require the Python Pyro module. Please install it."
    sys.exit(1)


from satellite import Satellite
from util import to_int, to_bool
from module import Module, Modules

VERSION = "0.1"


#Our main APP class
class Poller (Satellite):
	do_checks = True #I do checks
	do_actions = False #but no actions
	#default_port = 7771
	
	properties = {
		'workdir' : {'default' : '/usr/local/shinken/src/var', 'pythonize' : None, 'path' : True},
		'pidfile' : {'default' : '/usr/local/shinken/src/var/pollerd.pid', 'pythonize' : None, 'path' : True},
		'port' : {'default' : '7771', 'pythonize' : to_int},
		'host' : {'default' : '0.0.0.0', 'pythonize' : None},
		'user' : {'default' : 'shinken', 'pythonize' : None},
		'group' : {'default' : 'shinken', 'pythonize' : None},
		'idontcareaboutsecurity' : {'default' : '0', 'pythonize' : to_bool}
		}


################### Process launch part
def usage(name):
    print "Shinken Poller Daemon, version %s, from : " % VERSION
    print "        Gabes Jean, naparuba@gmail.com"
    print "        Gerhard Lausser, Gerhard.Lausser@consol.de"
    print "Usage: %s [options] [-c configfile]" % name
    print "Options:"
    print " -c, --config"
    print "\tConfig file."
    print " -d, --daemon"
    print "\tRun in daemon mode"
    print " -r, --replace"
    print "\tReplace previous running scheduler"
    print " -h, --help"
    print "\tPrint detailed help screen"
    print " --debug"
    print "\tDebug File. Default : no use (why debug a bug free program? :) )"


#lets go to the party
if __name__ == "__main__":
    #Manage the options
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hrdc::w", ["help", "replace", "daemon", "config=", "debug=", "easter"])
    except getopt.GetoptError, err:
        # print help information and exit:
        print str(err) # will print something like "option -a not recognized"
        usage(sys.argv[0])
        sys.exit(2)
    #Default params
    config_file = None
    is_daemon=False
    do_replace=False
    debug=False
    debug_file=None
    for o, a in opts:
        if o in ("-h", "--help"):
            usage(sys.argv[0])
            sys.exit()
	elif o in ("-r", "--replace"):
            do_replace = True
        elif o in ("-c", "--config"):
            config_file = a
        elif o in ("-d", "--daemon"):
            is_daemon = True
	elif o in ("--debug"):
            debug = True
	    debug_file = a
        else:
            print "Sorry, the option",o, a, "is unknown"
	    usage(sys.argv[0])
            sys.exit()

    p = Poller(config_file, is_daemon, do_replace, debug, debug_file)
    #import cProfile
    p.main()
    #command = """p.main()"""
    #cProfile.runctx( command, globals(), locals(), filename="/tmp/Poller.profile" )

